bundle/Service/ScheduledVisibilityService.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Netgen\Bundle\IbexaScheduledVisibilityBundle\Service;
+
+use DateTime;
+use Ibexa\Contracts\Core\Repository\Values\Content\Content;
+use Ibexa\Contracts\Core\Repository\Values\Content\Field;
+use Ibexa\Core\FieldType\Date\Value as DateValue;
+use Ibexa\Core\FieldType\DateAndTime\Value as DateAndTimeValue;
+use Netgen\Bundle\IbexaScheduledVisibilityBundle\Configuration\ScheduledVisibilityConfiguration;
+use Netgen\Bundle\IbexaScheduledVisibilityBundle\Enums\VisibilityUpdateResult;
+use Netgen\Bundle\IbexaScheduledVisibilityBundle\Exception\InvalidStateException;
+use Netgen\Bundle\IbexaScheduledVisibilityBundle\ScheduledVisibility\Registry;
+
+final class ScheduledVisibilityService
+{
+    public function __construct(
+        private readonly ScheduledVisibilityConfiguration $configurationService,
+        private readonly Registry $registry,
+    ) {}
+
+    /**
+     * @throws InvalidStateException
+     */
+    public function updateVisibilityIfNeeded(Content $content, ?string $handlerIdentifier = null): VisibilityUpdateResult
+    {
+        if (!$this->isValid($content)) {
+            throw new InvalidStateException($content);
+        }
+
+        $handler = $this->registry->get($handlerIdentifier ?? $this->configurationService->getHandlerIdentifier());
+
+        if ($this->shouldBeHidden($content) && !$handler->isHidden($content)) {
+            $handler->hide($content);
+
+            return VisibilityUpdateResult::Hidden;
+        }
+
+        if ($this->shouldBeVisible($content) && !$handler->isVisible($content)) {
+            $handler->reveal($content);
+
+            return VisibilityUpdateResult::Revealed;
+        }
+
+        return VisibilityUpdateResult::NoChange;
+    }
+
+    public function shouldBeHidden(Content $content): bool
+    {
+        $publishFromField = $content->getField('publish_from');
+        $publishToField = $content->getField('publish_to');
+
+        $publishFromDateTime = $this->getDateTime($publishFromField);
+        $publishToDateTime = $this->getDateTime($publishToField);
+
+        if ($publishToDateTime === null && $publishFromDateTime === null) {
+            return false;
+        }
+
+        $currentDateTime = new DateTime();
+
+        if ($publishFromDateTime !== null && $publishFromDateTime > $currentDateTime) {
+            return true;
+        }
+
+        if ($publishToDateTime !== null && $publishToDateTime <= $currentDateTime) {
+            return true;
+        }
+
+        return false;
+    }
+
+    public function shouldBeVisible(Content $content): bool
+    {
+        $publishFromField = $content->getField('publish_from');
```

Comment:
Here and in `shouldBeHidden()` add `isValid()` check and throw `InvalidStateException` exception. Then we can also type hint that the returned fields are an instance of `Field` and not null.

Reply:
fixed in 048fdf98159d3f1837501a77577acdfef4c312d0